src/cc/lib/graph/partition.cc

```diff
@@ -908,16 +908,38 @@ std::optional<size_t> Partition::EdgeFeatureOffset(uint64_t internal_src_node_id
     {
         return std::nullopt;
     }
+
     const auto tp_count = m_edge_type_offset[type_offset + 1] - m_edge_type_offset[type_offset];
     auto fst = std::begin(m_edge_destination) + m_edge_type_offset[type_offset];
     auto lst = fst + tp_count;
-    auto it = std::lower_bound(fst, lst, input_edge_dst);
-    if (it == lst)
+
+    // Fast path: linear search outperforms binary search for small number of elements.
+    if (tp_count <= 16)
     {
-        // Edge was not found in this partition.
+        auto it = std::find(fst, lst, input_edge_dst);
+        if (it != lst && *it == input_edge_dst)
+        {
+            return {it - std::begin(m_edge_destination)};
+        }
+
         return std::nullopt;
     }
-    return {it - std::begin(m_edge_destination)};
+
+    auto it = std::lower_bound(fst, lst, input_edge_dst);
```

Comment:
I don't see any validation that m_edge_destination is sorted per lower_bound requirements - at least not in the C++ code, I'm sure in python converter it is, but should we have such verification if algo depends on it?

Reply:
We don’t need this validation. Plain fallback to a linear search when element not found works.